File: redhawk/src/base/framework/java/ossie/src/org/ossie/logging/RH_LogEventAppender.java
/*
 * This file is protected by Copyright. Please refer to the COPYRIGHT file
 * distributed with this source distribution.
 *
 * This file is part of REDHAWK core.
 *
 * REDHAWK core is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * REDHAWK core is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses/.
 */

package org.ossie.logging;

import java.lang.Integer;
import org.omg.CORBA.Any;
import org.omg.CORBA.ORB;

import org.apache.log4j.AppenderSkeleton;
import org.apache.log4j.helpers.LogLog;
import org.apache.log4j.spi.Filter;
import org.apache.log4j.Layout;
import org.apache.log4j.spi.LoggingEvent;
import org.ossie.events.*;
import org.ossie.events.Manager.*;
import CF.LogEvent;
import CF.LogEventHelper;
import CF.LogEventHolder;
import org.ossie.component.Resource;

public class RH_LogEventAppender extends AppenderSkeleton  {

    //
    // Command line arguments used to configure orb
    //
    private String                                              _args;

    //
    // reconnection limits and retries...
    //
    private int                                                 _reconnect_retries=10;
    private int                                                 _reconnect_delay=10;

    //
    // channel name
    //
    private String                                              channelName;

    //
    // naming context 
    //
    private String                                              nameContext;

    //
    // Producer Identifier
    //
    private String                                              prodId = "RESOURCE.ID";

    //
    // Producer Name
    //
    private String                                              prodName="RESOURCE.NAME";

    //
    // Producer FQN - fully qualified domain name for resource
    //
    private String                                              prodFQN="RESOURCE.FQN";

    //
    // channel name, shadow variable
    //
    private String                                              _channelName;

    //
    // naming context, shadow variable
    //
    private String                                              _nameContext;

    //
    // Event Channel Manager that manages those resources for the Domain
    //
    private Manager                                             _ecm=null;
    
    //
    // Interface use to publish LogEvent messages to consumers
    //
    private Publisher                                           _pub=null; 

    // clean up event channel when appender is removed
    private int                                                 _cleanup_event_channel=0;

    // sychronization object
    private Object                                              updatingLock = null;

    public RH_LogEventAppender(){
        super();
        this.updatingLock = new Object();
    }

    public void activateOptions() {
        synchronized(this.updatingLock){
            if ( _channelName != channelName && channelName != "" ) {
                _channelName = channelName;
                _nameContext = nameContext;
                LogLog.debug("activate options :" + channelName );
                connect();
            }
            super.activateOptions();
        }
    }

    public void append(final LoggingEvent logRecord){
        synchronized(this.updatingLock){
            if (this.layout != null){

                if ( _pub != null ) {
                    LogLog.debug("Publish event to channel :" + channelName );
                    try {
                        final Any any = ORB.init().create_any();
                        int eventLogLevel = logging.ConvertLog4ToCFLevel(logRecord.getLevel());
                        long now =  System.currentTimeMillis()/1000;
                        String msg = layout.format(logRecord);
                        CF.LogEvent evt = new CF.LogEvent( prodId,
                                                           prodName,
                                                           prodFQN,
                                                           now,
                                                           eventLogLevel,
                                                           msg );
                        CF.LogEventHelper.insert( any, evt);
                        if ( _pub.push( any ) != 0 ) {
                            errorHandler.error("Unable to push message to channel:" + channelName );
                        }
                    }
                    catch( Throwable e ) {
                        errorHandler.error("Unable to push message to channel:" + channelName );
                    }
                
                }
                else {
                    if( _ecm != null ) {
                        errorHandler.error("No Publisher available for channel:" + channelName );
                    }
                }
            }
            else {
                errorHandler.error("No Layout set for the appender name [ " + name + " ].");
            }
        }
        return;
    }

    //
    // perform connect operation to establish a corba context 
    //
    public int setEventChannelManager( org.ossie.events.Manager ECM ) {
        _ecm = ECM;
        // if we are not connected then try to connect
        if ( _pub == null ) {
            LogLog.debug("Setting event channel manager  channel:" + channelName );
            return connect();
        }
        return 0;
    }

    //
    // perform connect operation to establish a corba context 
    //
    public int connect() {
        return connect(_ecm);
    }

    public int connect( org.ossie.events.Manager ECM ) {
        int retval = 0;

        LogLog.debug("Trying to Established Publisher for Channel:" + _channelName );
        if ( ECM != null )  {
            retval=-1;
            try {
                if ( _channelName != null && !_channelName.equals("") ) {
                    _pub = ECM.Publisher(_channelName);
                    retval=0;
                    LogLog.debug("Established Publisher for Channel:" + _channelName );
                }
            }
            catch( Manager.RegistrationExists e ){
            }
            catch( Manager.RegistrationFailed e ){
            }
        }
        else {
            /** RESOLVE 
                Get Event channel from org.ossie.corba.utils - getPublisher...
            **/
        }

        return retval;
    }

	 
    public void close()
    {
        LogLog.debug( "RH_LogEventAppender::close START");
        if ( closed ) return;
        
        if ( _pub != null ) {
            _pub.terminate();
        }
        _pub = null;

        
        closed=true;
        LogLog.debug( "RH_LogEventAppender::close END");
    }

    public boolean requiresLayout(){
        return true;
    }

    // set/get for event_channel or EVENT_CHANNEL
    public void setevent_channel(String value){
        synchronized(this.updatingLock){
            this.channelName = value;
        }
    }

    public String getevent_channel(){
        return this.channelName;
    }

    // set/get for name_context or NAME_CONTEXT
    public void setname_context(String value){
        synchronized(this.updatingLock){
            this.nameContext = value;
        }
    }

    public String getname_context(){
        return this.nameContext;
    }


    // set/get for producer_id or PRODUCER_ID
    public void setproducer_id(String value){
        synchronized(this.updatingLock){
            this.prodId = value;
        }
    }

    public String getproducer_id(){
        return this.prodId;
    }

    // set/get for producer_name or PRODUCER_NAME
    public void setproducer_name(String value){
        synchronized(this.updatingLock){
            this.prodName = value;
        }
    }

    public String getproducer_name(){
        return this.prodName;
    }


    // set/get for producer_fqn or PRODUCER_FQN
    public void setproducer_fqn(String value){
        synchronized(this.updatingLock){
            this.prodFQN = value;
        }
    }

    public String getproducer_fqn(){
        return this.prodFQN;
    }

    // set/get for argv or ARGV
    public void setargv(String value){
        synchronized(this.updatingLock){
            System.out.println("argv: " + value);
            this._args = value;
        }
    }

    public String getargv(){
        return this._args;
    }

    // set/get for remove_on_destroy or REMOVE_ON_DESTROY
    public void setremove_on_destroy(String value){
        synchronized(this.updatingLock){
            int nds = Integer.parseInt(value);
            if (nds == 0 || nds == 1){
                this._cleanup_event_channel = nds;
            }
        }
    }

    public int getremove_on_destroy(){
        return this._cleanup_event_channel;
    }

    // set/get for retries or RETRIES
    public void setretries(String value){
        synchronized(this.updatingLock){
            this._reconnect_retries = Integer.parseInt(value);
        }
    }

    public int getretries(){
        return this._reconnect_retries;
    }

    // set/get for retry_delay or RETRY_DELAY
    public void setretry_delay(String value){
        synchronized(this.updatingLock){
            this._reconnect_delay = Integer.parseInt(value);
        }
    }

    public int getretry_delay(){
        return this._reconnect_delay;
    }

}

File: redhawk/src/base/framework/logging/RH_LogEventAppender.cpp
/*
 * This file is protected by Copyright. Please refer to the COPYRIGHT file
 * distributed with this source distribution.
 *
 * This file is part of REDHAWK core.
 *
 * REDHAWK core is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * REDHAWK core is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses/.
 */
#ifdef   HAVE_LOG4CXX
#include <iostream>
#include <ossie/CF/LogInterfaces.h>
#include "RH_LogEventAppender.h"
#include <ossie/CorbaUtils.h>
#include <log4cxx/helpers/loglog.h>
#include <log4cxx/helpers/optionconverter.h>
#include <log4cxx/helpers/synchronized.h>
#include <log4cxx/helpers/transcoder.h>
#include <log4cxx/helpers/stringhelper.h>

using namespace log4cxx;
using namespace log4cxx::helpers;

#define _LL_DEBUG( msg ) \
  { std::ostringstream __os; __os << msg; LogLog::debug(__os.str()); __os.str(""); }

#define _LLS_DEBUG( os, msg ) \
  os << msg; LogLog::debug(os.str()); os.str("");


struct OrbContext;
typedef OrbContext*  OrbPtr;
 
//
// Orb
//
// Context for access to ORB and common CORBA services
//
struct OrbContext {

  // orb instantiation
  CORBA::ORB_ptr                          orb;

  // root POA for to handle object requests
  PortableServer::POA_ptr                 rootPOA;

  // handle to naming service
  CosNaming::NamingContext_ptr            namingService;

  // handle to naming service
  CosNaming::NamingContextExt_ptr         namingServiceCtx;

  virtual ~OrbContext() {};

  OrbContext() {
    orb = ossie::corba::Orb();
    rootPOA = ossie::corba::RootPOA();
    namingService = ossie::corba::InitialNamingContext();
    namingServiceCtx = CosNaming::NamingContextExt::_nil();
    try {
      CORBA::Object_ptr obj;
      obj=orb->resolve_initial_references("NameService");
      namingServiceCtx = CosNaming::NamingContextExt::_narrow(obj);
    }
    catch(...){
    };
  };

};




namespace log4cxx {
  // 
  // Allows for same naming reference to LogEvent Appender class
  //
    class ClassRH_LogEventAppender : public Class 
    {
    public:
        ClassRH_LogEventAppender() : helpers::Class() {}
        virtual LogString getName() const {
            return LOG4CXX_STR("org.ossie.logging.RH_LogEventAppender");
        }
        virtual ObjectPtr newInstance() const {
          return new RH_LogEventAppender();
        }
    };
}


// Register this class with log4cxx
IMPLEMENT_LOG4CXX_OBJECT_WITH_CUSTOM_CLASS(RH_LogEventAppender, ClassRH_LogEventAppender)


RH_LogEventAppender::RH_LogEventAppender():
channelName("LOG_CHANNEL"),
	nameContext(""),
	prodId("RESOURCE.ID"),
	prodName("RESOURCE.Name"),
	prodFQN("RESOURCE.FQN"),
	_channelName(""),
	_nameContext(""),
	_reconnect_retries(10),
	_reconnect_delay(10),
	_cleanup_event_channel(0)
{

}


RH_LogEventAppender::~RH_LogEventAppender() {

  if ( _event_channel &&  _cleanup_event_channel ) {
      _event_channel.reset();
      ossie::events::DeleteEventChannel( _channelName, _nameContext );
  }

}


void RH_LogEventAppender::setOption(const LogString& option, const LogString& value) {

    if(StringHelper::equalsIgnoreCase(option, LOG4CXX_STR("EVENT_CHANNEL"), LOG4CXX_STR("event_channel"))) {
      synchronized sync(mutex);
      channelName = value;
      _LL_DEBUG("RH_LogEventAppender::setOption event_channel: " << value );
    }
    else if(StringHelper::equalsIgnoreCase(option, LOG4CXX_STR("NAME_CONTEXT"), LOG4CXX_STR("name_context"))) {
      synchronized sync(mutex);
      nameContext = value;
      _LL_DEBUG("RH_LogEventAppender::setOption name_context: " << value );
    }
    else if(StringHelper::equalsIgnoreCase(option, LOG4CXX_STR("PRODUCER_ID"), LOG4CXX_STR("producer_id"))) {
      synchronized sync(mutex);
      prodId = value;
      _LL_DEBUG("RH_LogEventAppender::setOption producer_id: " << value );
    }
    else if(StringHelper::equalsIgnoreCase(option, LOG4CXX_STR("PRODUCER_NAME"), LOG4CXX_STR("producer_name"))) {
      synchronized sync(mutex);
      prodName = value;
      _LL_DEBUG("RH_LogEventAppender::setOption producer_name: " << value );
    }
    else if(StringHelper::equalsIgnoreCase(option, LOG4CXX_STR("PRODUCER_FQN"), LOG4CXX_STR("producer_fqn"))) {
      synchronized sync(mutex);
      prodFQN = value;
      _LL_DEBUG("RH_LogEventAppender::setOption producer_fqn: " << value );
    }
    else if(StringHelper::equalsIgnoreCase(option, LOG4CXX_STR("ARGV"), LOG4CXX_STR("argv"))) {
      synchronized sync(mutex);
      _args = value;
      _LL_DEBUG("RH_LogEventAppender::setOption argv: " << value );
    }
    else if(StringHelper::equalsIgnoreCase(option, LOG4CXX_STR("REMOVE_ON_DESTROY"), LOG4CXX_STR("remove_on_destroy"))) {
      synchronized sync(mutex);
      int nds = StringHelper::toInt(value);
      if ( nds == 0 || nds == 1 ) {
	_cleanup_event_channel = nds;
      }
    }
    else if(StringHelper::equalsIgnoreCase(option, LOG4CXX_STR("RETRIES"), LOG4CXX_STR("retries"))) {
      synchronized sync(mutex);
      int newRetry = StringHelper::toInt(value);
      _LL_DEBUG("RH_LogEventAppender::setOption retries: " << value );
      if ( newRetry > 0 ) {
	_reconnect_retries = newRetry;
      }
    }
    else if(StringHelper::equalsIgnoreCase(option, LOG4CXX_STR("RETRY_DELAY"), LOG4CXX_STR("retry_delay"))) {
      synchronized sync(mutex);
      int newDelay = StringHelper::toInt(value);
      _LL_DEBUG("RH_LogEventAppender::setOption retry_delay: " << value );
      if ( newDelay > 0 ) {
	_reconnect_delay = newDelay;
      }
    }
    else {
      _LL_DEBUG("RH_LogEventAppender::setOption non-appender option: value : " << value );
      AppenderSkeleton::setOption(option, value);
    }
}


void RH_LogEventAppender::activateOptions(Pool& p) {

  synchronized sync(mutex);
  std::ostringstream os;
  _LLS_DEBUG( os, "RH_LogEventAppender: CH:" << channelName ); 
  _LLS_DEBUG( os, "RH_LogEventAppender: NameContext:" << nameContext );
  _LLS_DEBUG( os, "RH_LogEventAppender: Retries:" << _reconnect_retries);
  _LLS_DEBUG( os, "RH_LogEventAppender: RetryDelay:" << _reconnect_delay);

  if ( _channelName != channelName && channelName != "" ) {
    LOG4CXX_ENCODE_CHAR(t, channelName );
    _channelName = t;
    LOG4CXX_ENCODE_CHAR(t2, nameContext );
    _nameContext = t2;
    connect_();
  }

  AppenderSkeleton::activateOptions(p);
  
}

	 
void RH_LogEventAppender::append(const spi::LoggingEventPtr& event, Pool& p){
  if ( this->layout == NULL ) {
    LOG4CXX_ENCODE_CHAR(nameStr, name);
    std::string msg("No Layout set for the appender named [ ");
    msg.append(nameStr);
    msg.append(" ].");
	 
    LOG4CXX_DECODE_CHAR(msgL, msg);
    errorHandler->error(msgL);
    return;
  }
	 
  log4cxx::LogString fMsg;
	 
  this->layout->format(fMsg, event, p);
	 
  LOG4CXX_ENCODE_CHAR(fMsgStr, fMsg);

  // This is the message structure for a Redhawk logging event
  CF::LogEvent rh_event;
  LOG4CXX_ENCODE_CHAR(t1,prodId);
  rh_event.producerId = CORBA::string_dup(t1.c_str());
  LOG4CXX_ENCODE_CHAR(t2,prodName);
  rh_event.producerName = CORBA::string_dup(t2.c_str());
  LOG4CXX_ENCODE_CHAR(t3,prodFQN);
  rh_event.producerName_fqn = CORBA::string_dup(t3.c_str());

  CORBA::Long level=CF::LogLevels::FATAL;
  if ( event->getLevel() == log4cxx::Level::getError() )
    level=CF::LogLevels::ERROR;
  if ( event->getLevel() == log4cxx::Level::getWarn() )
    level=CF::LogLevels::WARN;
  if ( event->getLevel() == log4cxx::Level::getInfo() )
    level=CF::LogLevels::INFO;
  if ( event->getLevel() == log4cxx::Level::getDebug() )
    level=CF::LogLevels::DEBUG;
  if ( event->getLevel() == log4cxx::Level::getTrace() )
    level=CF::LogLevels::TRACE;
  if ( event->getLevel() == log4cxx::Level::getAll() )
    level=CF::LogLevels::ALL;
  rh_event.level = level;

  //timeStamp in LoggingEventPtr is in microseconds
  //need to convert to seconds for rh_event
  rh_event.timeStamp = event->getTimeStamp()/1000000;
  rh_event.msg = CORBA::string_dup(fMsg.c_str());
  
  // push log message to the event channel
  if ( _event_channel ) {
    if ( _event_channel->push(rh_event) != 0 ) {
      _LL_DEBUG( "RH_LogEventAppender::append EVENT CHANNEL, PUSH OPERATION FAILED.");
    }
 }
  
}
	 
void RH_LogEventAppender::close()
{
  _LL_DEBUG( "RH_LogEventAppender::close START");
  if ( closed ) return;
  _event_channel.reset();
  closed=true;
  _LL_DEBUG( "RH_LogEventAppender::close END");
}


int RH_LogEventAppender::connect_() {

  int retval = 0;

  _event_channel.reset();
  std::ostringstream os;
  _LLS_DEBUG( os, "RH_LogEventAppender::connect Create PushEventSupplier" << _channelName );
  ossie::events::PushEventSupplier *pes=new ossie::events::PushEventSupplier( _channelName, 
							      _nameContext, 
							      _reconnect_retries, 
							      _reconnect_delay );
  if (pes != NULL ) {
    _LLS_DEBUG( os, "RH_LogEventAppender::connect Create PushEventSupplier Created." );
    _event_channel.reset(pes);
  }

  return retval;
}

#endif   //   HAVE_LOG4CXX
